google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatchingPublisher.java

```diff
@@ -0,0 +1,209 @@
+    private final RoutingPolicy routingPolicy;
+
+    private PartitionsWithRouting(
+        ImmutableMap<Partition, Publisher<PublishMetadata>> publishers,
+        RoutingPolicy routingPolicy) {
+      this.publishers = publishers;
+      this.routingPolicy = routingPolicy;
+    }
+
+    public ApiFuture<PublishMetadata> publish(Message message) throws CheckedApiException {
+      try {
+        Partition routedPartition =
+            message.key().isEmpty()
+                ? routingPolicy.routeWithoutKey()
+                : routingPolicy.route(message.key());
+        checkState(
+            publishers.containsKey(routedPartition),
+            String.format(
+                "Routed to partition %s for which there is no publisher available.",
+                routedPartition));
+        return publishers.get(routedPartition).publish(message);
+      } catch (Throwable t) {
+        throw toCanonical(t);
+      }
+    }
+
+    public void flush() throws IOException {
+      for (Publisher<PublishMetadata> publisher : publishers.values()) {
+        publisher.flush();
+      }
+    }
+
+    public void stop() {
+      publishers.values().forEach(ApiService::stopAsync);
+      publishers.values().forEach(ApiService::awaitTerminated);
+    }
+  }
+
+  private final CloseableMonitor monitor = new CloseableMonitor();
+
+  @GuardedBy("monitor.monitor")
+  private boolean shutdown = false;
+
+  @GuardedBy("monitor.monitor")
+  private Optional<PartitionsWithRouting> partitionsWithRouting = Optional.empty();
+
+  public PartitionCountWatchingPublisher(
+      PartitionCountWatcher.Factory watcherFactory,
+      PartitionPublisherFactory publisherFactory,
+      RoutingPolicy.Factory policyFactory) {
+    this.publisherFactory = publisherFactory;
+    this.policyFactory = policyFactory;
+    PartitionCountWatcher configWatcher = watcherFactory.newWatcher(this::handleConfig);
+    addServices(configWatcher);
+  }
+
+  @Override
+  public ApiFuture<PublishMetadata> publish(Message message) {
+    Optional<PartitionsWithRouting> partitions;
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      partitions = partitionsWithRouting;
+    }
+    if (!partitions.isPresent()) {
+      throw new IllegalStateException("Publish called before start or after shutdown");
+    }
+    try {
+      return partitions.get().publish(message);
+    } catch (CheckedApiException e) {
+      onPermanentError(e);
+      return ApiFutures.immediateFailedFuture(e);
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    Optional<PartitionsWithRouting> partitions;
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      partitions = partitionsWithRouting;
+    }
+    if (!partitions.isPresent()) {
+      throw new IllegalStateException("Publish called before start or after shutdown");
+    }
+    partitions.get().flush();
+  }
+
+  private ImmutableMap<Partition, Publisher<PublishMetadata>> addNewPartitions(
```

Comment:
This doesn't really "add". Call this "getNewPartitionPublishers"?

Reply:
Done